content/blog/2023-remote-data-101/index.mdx

```diff
@@ -0,0 +1,145 @@
+
+One of the key strategies in building a reliable app is to understand what data is flowing through your system. Unless you work in a small team that believes in _full-stack development_, you won't have much control over what comes to your app from the backend services. This data can come with multiple types of issues with a varying impact:
+
+* It can simply be invalid for your app. Let's say you have a B2C ecommerce app and you get a product without a name or price. While (for various legacy reasons) this might be valid on the backend it definitely is not a product that you can display like all the others. Either you'll have to build a special version of all the product details screens and listings to accomodate a product like that, or you just ignore it.
+* Data can be malformed. An ID that is a number is passed as a string, or URLs for any detailed pages are not localised. Those are issues that can (and have to!) be corrected _somewhere_ in your app.
+* Cosmetic issues: typos in property names, mixing snake case and camel case, misleading naming (field called `distanceInMiles` that returns kilometers instead), etc. Those issues can be fixed on your side, or accepted as quirks and propagated across your codebase as well.
+
+Of course, a lot of those issues are simply differences between how your team sees the domain (and how things _should_ be) and how other teams do. Sometimes there's a business logic error hidden there that you can simply ask the other team to fix, but a lot of times it's a matter of preference. Of course, even if you're right it doesn't mean that the problem is worth fixing. Good luck [raising priority on your ticket that somebody misspelled _referrer_](https://en.wikipedia.org/wiki/HTTP_referer).
+
+If you value maintainability in the long-term over moving fast and breaking things, you'll probably want to draw a line where the data enters your app and put a border guard there that does two things:
+* Checks for correctness
+* Translates from external (backend) to internal (yours) representation (and back to external format when data's leaving your app)
+
+You _could_ hand-craft a layer of artisanal code somewhere between low-level networking and your business logic, but you don't have to. Enter [zod](https://zod.dev/).
+
+## Data validation
+
+Let's take as an example the Marvel API's endpoint for characters. We'd like to represent a list of them with a possibility of fetching more data for a detailed screen should our user tap any of them. We _could_ do manual checks for the fields that we need (`ID` - to later fetch the details, `name` - to display it in our list, `URL` - to fetch a thumbnail image, and `modified` time - to possibly help us with cache invalidation), but we can also let `zod` do it for us by defining a schema that we're interested in:
```

Comment:
> We'd like to represent a list of them with a possibility of fetching more data for a detailed screen should our user tap any of them.

Is it grammatically correct? Looks a bit weird, but I guess it is. You know English better than me